### PR TITLE
Follow synonyms in #show_module _type

### PR DESCRIPTION
**Pull request by gasche**
This PR against 4.14 is intended as a fix for #11533.

At the time of writing, it contains ad-hoc logic in the implementation of `#show_module_type` to follow a chain of synonyms, mirroring [similar logic](https://github.com/ocaml/ocaml/blob/4.14/toplevel/topdirs.ml#L547) in the implementation of `#show_module` to follow a chain of aliases.

Fixes: #11533

### CODE REVIEW

**Comment by gasche** [comment]
(Reviews are already welcome.)

**Comment by gasche** [comment]
I added some tests.

While playing with `#show` in presence of module aliases or module type synonyms I encountered some strange behaviors

```ocaml
# module U = Stdlib.Unit;;
module U = Unit
# #show U;;
module U = Unit
module U = Unit
module U :
  sig
    type t = unit = ()
    val equal : t -> t -> bool
    val compare : t -> t -> int
    val to_string : t -> string
  end
```

Note the repeated line `module U = Unit` in the `#show U` output, which is not part of the regression reported in #11533 -- but probably also an interaction with implicit strengthening during environment lookup. After the present PR, the same issue can now show up for module types as well.

This stuttering is independent of the present regression, I am not sure how to fix it, so I propose to leave it outside the scope of the present bugfix PR. (It is also a lesser problem, as it adds some useless noise in the output but does not hide the previously-useful output.)

**Comment by gasche** [comment]
This small PR is ripe for review. I you look at the code, it is useful to compare it with the code of `#show_module` above:

https://github.com/ocaml/ocaml/blob/adcf2cb2c6bea7701d399e7ec8734b55403584a3/toplevel/topdirs.ml#L539-L561

File: toplevel/topdirs.ml
Comment (trefis) [inline toplevel/topdirs.ml:586]: I'd be tempted to extract the `Some (Mty_alias _)` case, to explicitly state that it can't happen.

**Comment by xavierleroy** [comment]
Ping?  Who would be good reviewers for this PR?

File: toplevel/topdirs.ml
Comment (garrigue) [reply toplevel/topdirs.ml:586]: I'm not sure it would really help.
This is just a choice we made of not having module aliases in the syntax of module types, but this code would be correct anyway.

File: testsuite/tests/tool-toplevel/show.ml
Comment (garrigue) [inline testsuite/tests/tool-toplevel/show.ml]: `as little as possible`

File: testsuite/tests/tool-toplevel/show.ml
Comment (garrigue) [inline testsuite/tests/tool-toplevel/show.ml]: `so many`

File: toplevel/topdirs.ml
Comment (garrigue) [inline toplevel/topdirs.ml]: Couldn't you avoid stuttering by checking that you are not adding the same definition again?

**Review by garrigue — COMMENTED** [review]
This looks basically fine to me.
If fixing stuttering is easy, I would prefer it to be done, otherwise I do not care so much.

**Comment by Octachron** [comment]
I propose to fix the stuttering at a later date so we can have the fix in 4.14 and 5.0 (this is a toplevel fix, which I am willing to test in the release candidate).

**Comment by gasche** [comment]
Thanks for the ping! The increase in feature scope had somehow suspended this PR from my memory.

I integrated @garrigue's review comments, and I worked out a fix for the stuttering issue. (Guess what, it was caused by Stdlib__Unit. It's the second time in two weeks that this hidden prefixing adds some pain to my development life.)

**Comment by gasche** [comment]
This still needs an approval before we can merge.

**Review by Octachron — APPROVED** [review]
The code looks good to me (even if I would have fixed the stuttering in the reverse way).

**Comment by gasche** [comment]
By "the reverse way", do you mean that if the lookup chains finds first D1 and then D2, and in fact D2 double-underscore-simplifies into D1, you would rather keep D2 and throw away D1, rather than keep D1 and throw away D2 as I am doing? (That sounds slightly more difficult to implement for no obvious benefit.)

**Comment by Octachron** [comment]
As far as I can see, you are throwing away `D1` and keeping `D2` in your code since you only add a new item once you have confirmed that it would not be shadowed by a subsequent expansion: 
```ocaml
if secretly_the_same_path env path new_path
then acc
else def :: acc
```
?

But I agree that it would be more difficult with dubious benefits.

**Comment by gasche** [comment]
We are talking about two successive aliases that would become `module M = <path>` and `module M = <new_path>` in the expansion trace. If `path` and `new_path` (which is an expansion of `path`) are the same modulo double-underscore-rewrite, then the `def` corresponding to `module M = <new_path>` is omitted from the result, but `module M = <path>` is already in the `acc` variable. (Or maybe I'm the one reading things wrong, but whatever, both would be fine anyway.)

**Pull request merged**
